### ISSUE TITLE
feat(blcs,plcs): reference camera条件付きsymmetric CourtKP7 embeddingで推論座標系を一意化する

### ISSUE THREAD

**Issue by Motoki0705** [Motoki0705/tennis-lab#719]
## 概要

Court Detectionが出力する7個のsemantic keypoint classを、ordered CourtKP14へ復元せず、そのままBLCS / PLCSのmulti-view track-query modelへ入力できるembedding経路を実装します。

7 classでは各classにnear側・far側のphysical pointが同居し、各cameraの入力時点ではその対応が未確定です。本Issueでは、すべてのcameraへnear/far labelを与えるのではなく、**1つのcameraだけをreference viewとして指定**します。モデルはreference viewのgeometryを推論座標系の基準として受け取り、その他のviewについてはcourt geometry、object observation、multi-view対応、時系列情報からnear/far対応を潜在的に推論します。

最終的なper-detection observation tokenは次のshapeを正本とします。

```text
[B, V, T, D, H]
```

これを既存の`BLCSTrackQueryModel` / `PLCSTrackQueryModel`が持つper-time spatial fusionとper-query temporal fusionへ接続します。

## 背景

Synthetic Court datasetの7 classは次のsemanticを持ちます。

```text
0 doubles_left
1 doubles_right
2 singles_left
3 singles_right
4 service_left
5 service_right
6 service_t
```

各classにはnear/farの2点が同じclassとして含まれ、複数courtが写る場合はさらに複数peakが同じclassへ入ります。したがって、class内のpeak indexには固定された意味順序がありません。

一方、現行BLCS / PLCS track-query modelは次を前提とします。

```text
court_kp: [B, V, T, 14, 2]
court_vis: [B, V, T, 14]
```

- `CourtBallGroupEmbedding` / `CourtPlayerGroupEmbedding`はordered 14 pointをflattenする。
- `CourtBallPointFusion`は固定14 pointを個別token化し、point indexをRoPE座標として使用する。
- `BLCSTrackQueryModel` / `PLCSTrackQueryModel`は`num_court_tokens = 14`を固定する。

この経路へ7 classを単純に7点として入れることはできません。また、7 classの各peakへ`E_near/far`を付与すると、全cameraのnear/far対応を入力側で解決済みにしてしまい、モデルが担うべきmulti-view orientation reasoningを消してしまいます。

## 固定する設計原則

1. **near/farを全cameraへ入力しない。**
   - `E_near/far`は使用しない。
   - 各other viewのpeakについて、near/far assignmentを前処理で確定しない。

2. **reference signalは1 viewだけへ与える。**
   - 各sample / clipで有効なreference viewは厳密に1つとする。
   - その他のviewはすべてcoordinate-anchor roleとして`other`である。
   - `reference/other`はcamera identityではなく、推論座標系を固定する役割である。

3. **7 class内のpeak集合はunorderedとして扱う。**
   - peak index `n`へlearned index embeddingやRoPEを与えない。
   - class内のpeak permutationで出力が変化しないset aggregationを使用する。

4. **ordered CourtKP14へ暗黙変換しない。**
   - homography等で14 pointへ復元して既存flatten経路へ戻す方式を本実装の正本にしない。
   - 7 classのmulti-peak semanticsをモデル境界まで保持する。

5. **最終observation tokenへ`E_type`を加えない。**
   - local encoderがcourt geometryとball/player observationを1つのobject-conditioned tokenへ融合する。
   - 外側のtrack-query backboneでは、既存のslot embeddingおよびslot/observation role RoPEが責任を持つ。
   - learnedなcourt/ball/player token-type embeddingを最終tokenへ追加しない。

6. **reference viewを変えるとtarget座標系も変える。**
   - reference signalだけを変更して同一targetを使用しない。
   - reference cameraから見たnear側を負のY、far側を正のYとする出力契約へ全targetを変換する。

## Tensor contract

記号を次で固定します。

```text
B: batch
V: camera views
T: frames
D: ball/player detections per camera-time
C: semantic court classes (= 7)
N: maximum peaks per class after padding
J: player joints
H: hidden dimension
```

Court peak入力:

```text
court_peak_uv:          [B, V, T, C, N, 2]
court_peak_score:       [B, V, T, C, N]
court_peak_covariance:  [B, V, T, C, N, 2, 2]
court_peak_valid:       [B, V, T, C, N]
```

Reference指定:

```text
reference_view_index:   [B]
# またはboundary内部で同値な
reference_view_mask:    [B, V]
```

要件:

- `reference_view_mask.sum(dim=1) == 1`
- reference viewは`view_mask=True`
- class順は上記7 class schemaと一致する
- `N`は可変peak数をpaddingするための軸であり、index自体に意味を持たない
- 複数courtのpeakをcourt-instance IDで事前分割しない
- covarianceを使用する場合はheatmap local moments等から明示的に生成する。利用できないsourceはsource adapterで明示した契約値を生成し、model内部で暗黙fallbackしない

BLCS object入力:

```text
ball_uv:          [B, V, T, D, 2]
ball_score:       [B, V, T, D]
ball_visible:     [B, V, T, D]
candidate_mask:   [B, V, T, D]
```

PLCS object入力:

```text
human_kp:         [B, V, T, D, J, 2]
joint_visibility: [B, V, T, D, J]
detection_score:  [B, V, T, D]
detection_mask:   [B, V, T, D]
```

### 1. Court peakとobject detectionのpair feature

Court peakを`p`、ballまたはplayer anchorを`a`とし、正規化相対座標を次で定義します。

```text
Δu = u_court - u_object
Δv = v_court - v_object
```

各court peakと各detectionの概念embeddingは次とします。

```text
e[b,v,t,d,c,n]
  = phi_pair(
      u_norm,
      v_norm,
      score,
      covariance,
      delta_u,
      delta_v
    )
  + E_class[c]
```

数式表現:

```text
e_{b,v,t,d,c,n}
  = φ_pair(u~, v~, s, Σ, Δu, Δv)
  + E_class(c)
```

ここには次を入れません。

```text
E_near/far
E_reference/other
E_type
E_peak_index(n)
```

`E_class`は7 semantic classだけを表し、near/far identityは表しません。

### 2. Unordered set aggregation

各detectionについて、全`(class, peak)`をmasked set aggregationし、1個のcamera-time-detection tokenへ変換します。

```text
g[b,v,t,d]
  = rho_set({e[b,v,t,d,c,n] | court_peak_valid})
```

出力:

```text
g: [B, V, T, D, H]
```

`rho_set`は次のいずれかの順序不変実装とします。

- object-queryからcourt peak setへのcross-attention
- masked attention pooling
- small Set Transformer

単純な`[C,N]` flatten + MLPを正本にしません。class内peak permutationと可変peak数に対して不安定になるためです。

### 3. Task-specific object geometry

BLCSではball UV自体がpairwise relative geometryへ含まれます。

PLCSでは、visibility-aware pelvis / pose center等の明示的anchorを`Δu, Δv`の基準にし、全身2D poseの形状情報をtask-specific object encoderで保持します。

概念上は次を許可します。

```text
g_blcs = rho_set(pair_features)
g_plcs = rho_set(pair_features) + psi_player(human_kp, joint_visibility)
```

ただし、最終出力はどちらも同じ`[B,V,T,D,H]` contractへ揃えます。

### 4. Reference / other signal

set aggregation後のview-level observation tokenへだけcamera role embeddingを加えます。

```text
z[b,v,t,d]
  = g[b,v,t,d]
  + E_view_role[reference_view_mask[b,v]]
```

数式表現:

```text
z_{b,v,t,d}
  = g_{b,v,t,d}
  + E_reference/other(r_{b,v})
```

ここで`r`は次の二値だけです。

```text
reference: 選択された1 camera
other:     それ以外の全camera
```

`E_reference/other`は`T`と`D`へbroadcastします。各other cameraへ個別のnear/far roleを与えません。

実装は2個のlearned vectorでも、`other=0`かつreferenceだけにlearned deltaを加える方式でも構いません。意味論上の正本は「referenceとotherの差分」です。

## 計算効率

上記数式をそのまま実装すると、`[B,V,T,D,C,N,H]`の大きな中間tensorを生成する可能性があります。概念contractは維持しつつ、production実装では以下を推奨します。

```text
court peaksを一度だけencode
  -> [B,V,T,C*N,h]

object detectionsをquery化
  -> [B,V,T,D,h]

relative geometryをattention bias / pair projectionとして計算

object-to-court cross-attention
  -> [B,V,T,D,H]
```

要件:

- H次元pair tensorの全面materializeを避けられる構造を優先する
- `C=7`、`N`、`D`に対するlatency / peak memoryを計測する
- 最適化のためにclass/peak semanticsをflatten固定順へ変更しない

## 推論座標系の定義

各sampleでreference cameraを1台選び、そのcameraから見たnear/farによりtarget座標系を一意化します。

source court座標の2つのbaseline midpointを次とします。

```text
B_plus  = (0, +HALF_LENGTH, 0)
B_minus = (0, -HALF_LENGTH, 0)
```

reference camera centerをcourt座標へ変換し、両baselineまでの距離を比較します。

```text
near baseline = reference cameraに近いbaseline
far baseline  = reference cameraから遠いbaseline
```

出力座標の規約:

```text
near side -> Y < 0
far side  -> Y > 0
```

source coordinateで`-Y`側がnearなら変換なし、`+Y`側がnearならYを反転します。

```text
x_ref = x_source
y_ref = orientation_sign * y_source
z_ref = z_source
```

reference cameraがcourtの真横に近く、両baseline距離差が閾値未満の場合は、orientationを推測せずreference候補から除外するか明示的に失敗させます。

### Target変換

reference orientationは次へ一貫して適用します。

BLCS:

- position
- velocityを出力する場合のY成分
- acceleration等のcourt-space vector
- physics / trajectory diagnosticsのcourt-space成分
- reprojection lossで使用する座標変換とcamera contract

PLCS:

- position
- heading / rotationのcourt-space direction
- court-space poseまたはjointを出力する場合のY成分
- canonical poseがplayer-localである場合は不要な反転を加えない
- visualization / metrics / serialization

rotationが`[cos(theta), sin(theta)]`でcourt XY方向を表す場合、Y反転時はdirection vectorのY成分を同じsignで変換します。角度値へ場当たり的な加減算を行わず、direction vectorとして変換してから正規化します。

## Track-query modelへの接続

最終token:

```text
z: [B, V, T, D, H]
```

既存backboneへ次の順で接続します。

```text
[B, V, T, D, H]
  -> permute
[B, T, V, D, H]
  -> flatten views and detections
[B, T, V*D, H]
  -> concatenate Q track slots
[B*T, Q + V*D, H]
  -> per-time spatial self-attention
  -> slot temporal self-attention [B*Q, T, H]
```

変更対象:

- `src/tasks/blcs/models/blcs_track_query_model.py`
- `src/tasks/plcs/models/plcs_track_query_model.py`
- 現行`CourtBallGroupEmbedding`
- 現行`CourtPlayerGroupEmbedding`
- 現行`CourtBallPointFusion`
- model I/O adapters、batch contracts、collate
- BLCS / PLCS target coordinate conversion

既存のslot/observation role RoPEは、learned`E_type`とは別責任なので維持可能です。`E_reference/other`は既存role axisへ混同せず、camera observation tokenのvalue streamへ加えます。

## Camera indexとreference roleの分離

現行spatial M-RoPEはcamera indexを持ちます。これはreference/other roleとは別概念です。

```text
camera index:
  viewを区別するためのspatial coordinate

reference/other:
  どのviewが出力座標系を固定するか
```

要件:

- reference cameraを常に`camera_0`へ固定しない
- 学習時にvalid viewからreferenceを選択し、targetを同時に変換する
- view順を固定shortcutとして利用しないよう、reference selectionと必要なview permutation augmentationを導入する
- camera M-RoPEを維持・無効化する比較を可能にしてもよいが、`camera index == reference identity`を学習させない

「other viewが等価」とはcoordinate-anchor roleが等価であることを意味し、各viewの観測geometryまで同一に潰すことではありません。

## Reference signalが消失しないこと

reference roleをper-detection tokenへ加えるため、reference viewで全detectionがmaskされるとsignalがspatial attentionから消える危険があります。

要件:

- `reference_view_mask`をdetection visibilityとは独立したbatch fieldとして保持する
- valid reference viewについて、少なくとも1つのobservation/context tokenがattentionへ残ることを保証する
- detectionが0件のcamera-timeを扱う場合は、D軸内に明示的なnull observationを確保する等、`[B,V,T,D,H]` contract内でreference signalを保持する
- invisible observationをmaskする設定でもreference signalが全clipから消えないことをテストする
- dedicated court-only tokenを追加する場合はshapeと責任変更を別途明示し、暗黙に`D+1`へ変更しない

## Data / I/O責務

Validationは`nn.Module.forward`へ置かず、#691後のdata / I/O boundaryで行います。

- class schema、shape、dtype、finite
- exactly-one reference view
- reference viewがview mask内に存在
- peak maskとpadding
- target orientation sign
- source targetとreference-oriented targetの整合
- BLCS / PLCS output fieldごとの変換漏れ

#714が定義する7 class multi-peak contractを入力正本として利用し、7 classからordered 14 pointへのcompatibility adapterを追加しません。

## Shared encoder責務

BLCS / PLCSでcourt set aggregationを二重実装しません。

概念構成:

```text
SymmetricCourtPeakEncoder
  - normalized peak geometry
  - score / covariance
  - E_class

CourtObjectSetFusion
  - object-relative geometry
  - masked set/cross attention
  - [B,V,T,D,H]

BLCS adapter
  - ball observation

PLCS adapter
  - player anchor + full pose feature

ReferenceViewConditioning
  - reference / other role
```

shared moduleはBLCS / PLCSの具体batch classやoutput headをimportせず、task adapter側がobject geometryを供給します。

### Court set encoder

- [ ] class内peak順をshuffleしてもtokenが許容誤差内で一致する
- [ ] class順の変更は対応するclass IDを同時に変更した場合に同値である
- [ ] invisible / padded peakが出力へ寄与しない
- [ ] 1 classに2点、4点以上、0点を含むbatchを処理できる
- [ ] 複数courtのpeakを同一classへ保持できる
- [ ] `E_near/far`、peak-index embedding、final `E_type`が存在しない
- [ ] output shapeが`[B,V,T,D,H]`

### Reference conditioning

- [ ] 各sampleでreference viewが厳密に1つである
- [ ] reference viewが`view_mask=False`ならboundaryで失敗する
- [ ] reference roleは選択viewの全T/Dへbroadcastされる
- [ ] other viewはすべて同じcamera-role embeddingを受ける
- [ ] reference indexを変えた場合、対応するtarget orientationも同時に変わる
- [ ] opposite-side referenceでY targetが反転し、X/Zが維持される
- [ ] side-on cameraを曖昧なまま採用しない
- [ ] detection欠損時もreference signalがbackboneへ残る

### BLCS

- [ ] variable `V/T/D/N`でforward / backwardできる
- [ ] candidate順shuffleに対してprediction集合が不変である
- [ ] position targetのY変換が正しい
- [ ] velocity等のoptional vector targetも同じorientation signを使う
- [ ] spatial / temporal stage間でshapeが維持される
- [ ] reference cameraを変えたcounterfactual batchを処理できる

### PLCS

- [ ] variable `V/T/D/N/J`でforward / backwardできる
- [ ] detection順shuffleに対してprediction集合が不変である
- [ ] position Yとheading directionの変換が整合する
- [ ] canonical poseへ不要なreflectionを加えない
- [ ] joint visibilityを含むplayer anchor生成がdeterministicである
- [ ] reference cameraを変えたcounterfactual batchを処理できる

### Integration

- [ ] 7 class court predictor / dataset outputからBLCS batchを構築できる
- [ ] 7 class court predictor / dataset outputからPLCS batchを構築できる
- [ ] ordered KP14への中間復元がactive pathに存在しない
- [ ] existing track-query spatial/temporal backboneを再利用している
- [ ] ruff、mypy、unit、integration、limited training checkが成功する

## 評価指標

BLCS:

- position error total
- X / Y / Z axis-wise error
- Y sign accuracy
- reference camera side別error
- visible view数別error
- peak欠損・multi-court条件別error
- presence / tracking metrics

PLCS:

- position error total
- X / Y / Z axis-wise error
- Y sign accuracy
- heading / rotation error
- reference camera side別error
- visible view数別error
- peak欠損・multi-court条件別error
- presence / tracking metrics

Reference consistency:

- 同一sceneでreferenceだけを変更したときのtarget transform consistency
- opposite-side referenceでのY / heading反転整合
- same-side reference間での出力安定性
- fixed camera index shortcutの有無

絶対閾値はbaseline測定前に恣意的に固定しません。ただし、`KP7 without reference`に対して`KP7 + reference`がY sign / Y errorを改善しない場合はproduction採用しません。

## Acceptance checklist

- [ ] Court Detectionの7 semantic multi-peak出力をordered KP14へ変換せずBLCS / PLCSへ入力できる
- [ ] per-peak embeddingがgeometryと`E_class`だけを持ち、`E_near/far`を持たない
- [ ] unordered peak setからper-detection token`[B,V,T,D,H]`を生成できる
- [ ] set aggregation後に1 viewだけへreference signalが入り、その他は同じother roleを持つ
- [ ] final observation tokenへlearned `E_type`を追加していない
- [ ] reference cameraに基づきBLCS / PLCS target Y軸が一意に定義される
- [ ] other cameraのnear/far assignmentを入力前処理で解決していない
- [ ] modelがexisting track-query spatial / temporal fusion経路を利用する
- [ ] class内peak順、detection順、reference view indexのshortcutに対するテストがある
- [ ] reference signalがdetection欠損やmaskで消失しない
- [ ] BLCS / PLCSのforward、backward、limited trainingが成功する
- [ ] `KP14 baseline`、`KP7 no-reference`、`KP7 reference-conditioned`の比較結果を記録する
- [ ] Y sign / Y error、PLCS heading、reference consistencyを定量評価する

## 非目標

- すべてのcameraへnear/far labelを付与すること
- court detectorにordered 14 pointを再び必須化すること
- 7 classからhomographyでKP14を復元して現行flatten embeddingへ戻すこと
- camera extrinsicsをmodel入力の必須条件にすること
- modelへ各other cameraのnear/farを明示出力させること
- Court Detection backbone / decoder自体の再設計
- track-query backbone全体の別architectureへの置換
- line / segmentation inputとの融合
- multi-court instance groupingを教師として追加すること

## 関連Issue

- #714 — 14 ordered pointと7 semantic multi-peakを扱うCourt data contract
- #695 — canonical synthetic scene / BLCS / PLCS dataset lifecycle
- #648 — Multi-ball BLCS track-query contract
- #643 — Multi-person PLCS track-query contract

## 完了条件

reference cameraを単にcamera index 0へ固定した実装、全cameraへnear/far embeddingを付けた実装、7 classを固定順7点としてflattenした実装、または7 classをKP14へ復元して既存modelへ渡す実装は未完了です。

1台だけのreference signal、unordered 7 class peak set、object-relative geometry、per-detection token`[B,V,T,D,H]`、reference-oriented target座標、既存track-query時空間融合が一つの整合した学習・推論contractとして成立し、BLCS / PLCSの比較runでY軸一意化の効果を確認できた場合に完了とします。
